README.md
# NAME

Mail::ToAPI::Checkvist - Email in tasks for Checkvist.com, see Mail2CV.com

# SYNOPSIS

    use Mail::ToAPI::Checkvist;

or

    % mail2cv.pl < email.eml

# DESCRIPTION

This is the script that powers Mail2CV.com.

It is a service that allows adding tasks to your Checkvist.com list via email.

You will need a fairly modern perl and some modules from CPAN:

uni::perl, WebService::Simple, Email::MIME, Email::Address

You will also need to set up some email forwarding to feed incoming
emails to the script.  I use a catch-all virtual domain in Postfix and
a simple ~/.forward with a pipe.

# AUTHOR

Alex Kapranoff <alex@kapranoff.ru>

# COPYRIGHT

Copyright 2013 Alex Kapranoff

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License, Version
3.

# SEE ALSO

[http://checkvist.com](http://checkvist.com)


Changes
Revision history for Mail-ToAPI-Checkvist

{{$NEXT}}

0.4  2013-07-28 17:46:36 MSK
        - Refactoring, notes, files, tests.
        The result of collaboration with the Checkvist.com team.


post2cv.pl
#!/usr/bin/perl

# Copyright 2013 Alex Kapranoff
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
# 
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

use uni::perl;
use WebService::Simple;
use Email::MIME;
use Email::Address;

open STDERR, '>>', '/home/kappa/2cv.log';

sub parse_email_from_stdin {
	my ($from, $to, $subject);

	binmode STDIN, ':bytes';
	my $email = Email::MIME->new(join "", <STDIN>);

	$subject = $email->header('Subject');
	$to   = (Email::Address->parse($email->header_obj->header_raw('To')))[0]->user;
	$from = (Email::Address->parse($email->header_obj->header_raw('From')))[0]->address;

	my ($remotekey, $list_id);
	if	($to =~ /^([^+]+)\+(\d+)$/) {
		($remotekey, $list_id) = ($1, $2);
	}
	elsif	($to =~ /^([^+]+)\+([^+]+)\+(\d+)$/) {
		($from, $remotekey, $list_id) = ($1, $2, $3);
		$from =~ tr/$=/@@/;
	}

	return ($from, $remotekey, $list_id, $subject);
}

sub add_task_to_checkvist {
	my ($login, $remotekey, $list_id, $task_text) = @_;
	my $chv = WebService::Simple->new(
	    base_url    => 'http://checkvist.com/',
	    response_parser => 'JSON',
	);
	$chv->credentials('checkvist.com:80', 'Application', $login, $remotekey);

	$chv->post("checklists/$list_id/tasks.json", {
		'task[content]' => $task_text,
	});
}

eval {
	say STDERR scalar localtime;
	my ($login, $remotekey, $list_id, $task_text)
		= parse_email_from_stdin();
	say STDERR "for $login";
	add_task_to_checkvist($login, $remotekey, $list_id, $task_text);
};

if ($@) {
	say STDERR 'Exception: ', $@;
}

exit 0;


script/mail2cv.pl
#!/usr/bin/perl

# Copyright 2013 Alex Kapranoff
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

use uni::perl;
use Mail::ToAPI::Checkvist;
use Config::Tiny;

my $Cfg = Config::Tiny->read("$ENV{HOME}/.mail2cvrc")
    || Config::Tiny->read('/etc/mail2cv/mail2cv.rc')
    || {};

$Mail::ToAPI::Checkvist::API_Endpoint = $Cfg->{checkvist}->{api_endpoint} // 'https://beta.checkvist.com';

open STDERR, '>>', $Cfg->{mail2cv}->{logfile} // "$ENV{HOME}/2cv.log";

eval {
    say STDERR scalar localtime;
    my $job = Mail::ToAPI::Checkvist->parse_email(*STDIN);

    say STDERR "for $job->{login}";
    Mail::ToAPI::Checkvist->execute($job);
};

if ($@) {
    say STDERR 'Exception: ', $@;
}

exit 0;


lib/Mail/ToAPI/Text.pm
package Mail::ToAPI::Text;

use strict;
use 5.010;
our $VERSION = '0.1';

use parent qw/Exporter/;
use HTML::Parser;
use HTML::Entities;
use List::Util qw/reduce first/;
use HTTP::Headers::Util;
use Encode;

our @EXPORT_OK = qw/_parse_for_content/;

sub _parse_header_fields {
    my $fields = { @{(HTTP::Headers::Util::split_header_words($_[0]))[0]} };

    # this is a parser for a subset of RFC2231
    # which is a big brother for RFC2047 MIME words
    # Mozilla Thunderbird uses RFC2231
    if (my $start = first { /\*0\*$/ } keys %$fields) {
        my $value = '';
        $start =~ /^([^*]+)\*0\*$/;
        my $field_name = $1;

        my $counter = 0;
        while (my $next_chunk = delete $fields->{"$field_name*$counter*"}) {
            $value .= $next_chunk;
            ++$counter;
        }
        $value =~ s/^([^']*)'([^']*)'//;
        my ($charset, $lang) = ($1, $2);

        $value =~ s/%([0-9A-Fa-f]{2})/chr(hex($1))/eg;

        $fields->{$field_name} = $charset
            ? decode($charset, $value)
            : $value;
    }

    return $fields;
}

sub _ct {
    my $data = _parse_header_fields("ct=$_[0]");

    return $data->{ct};
}

sub _may_contain_text {
    my $part = shift;

    my $ct = _parse_header_fields("ct=" . $part->content_type);

    return  $ct->{ct} =~ m{^multipart/}i
        || ($ct->{ct} =~ m{^text/(?:plain|html)$}i
            && ($part->header('Content-Disposition') || '') !~ /^attachment\b/i);
}

sub _render_recur {
    my $part = shift;
    my ($result_str, $files) = ('', undef);

    my $content_type = _ct($part->content_type) || 'text/plain';

    my $disp;
    if (my $cd = $part->header('Content-Disposition')) {
        $disp = _parse_header_fields("d=$cd");
    }

    if ($disp->{d} && $disp->{d} eq 'attachment') {
        push @$files, [$disp->{filename}, $content_type,
            $content_type =~ /^text\// ? $part->body_str : $part->body];
    }
    else {
        given ($content_type) {
            when ('multipart/related') {
                ($result_str, $files) = _render_recur(($part->subparts)[0]);
            }
            when ('multipart/alternative') {
                # choose the last of all supported, may also be
                # implemented as (grep)[-1]
                # if there's 1 subpart, always return it
                my $best_part = reduce { _may_contain_text($b) ? $b : $a } $part->subparts;
                ($result_str, $files) = _render_recur($best_part);
            }
            when (m{^multipart/}) { # 'mixed' and all others
                for my $subpart ($part->subparts) {
                    my ($part_str, $part_files) = _render_recur($subpart);
                    if (_may_contain_text($subpart)) {
                        $result_str .= $part_str . "\n";
                    }
                    push @$files, @$part_files if $part_files;
                }
                $result_str =~ s/\n\z//; # emulate join
            }
            when ('text/html') {
                $result_str = textify_html($part->body_str);
            }
            when (m{^text/}) {
                $result_str = $part->body_str;
            }
            default {
                # cannot render this part inline, so emulate attachment
                # should not happen very often
                push @$files, [$disp->{filename}, $content_type,
                    $content_type =~ /^text\// ? $part->body_str : $part->body];
            }
        }
    }

    return ($result_str, $files);
}

sub _parse_for_content {
    my $email = shift;

    my ($body_str, $files) = _render_recur($email);

    $body_str =~ s/^\s+//s;

    $body_str =~ s/(?:\r?\n|^)--[ ]?\r?\n.*\z//s;

    $body_str =~ s/\s+\z//s;

    return ($body_str, $files);
}

*unescapeHTMLFull = \&HTML::Entities::decode_entities;

sub textify_html($) {
    my $p = new HTML::Parser
        api_version => 3,
        start_h     => [\&_textify_tag, 'self, tagname, skipped_text, attr'],
        end_h       => [\&_textify_end_tag, 'self, tagname, skipped_text'],
        end_document_h=> [ sub { $_[0]->{result} .= _textify_decode($_[1]) }, 'self, skipped_text'],
        process_h   => [""],
        declaration_h=> [""],
        comment_h   => [""];

    $p->{result} = '';
    $p->{inside_bad_element} = undef;

    $p->parse($_[0]); # html (first arg)
    $p->eof;

    $p->{result} =~ s/[ \n\r\t]*\n[ \n\r\t]*\n/\n\n/gs;

    $p->{result} =~ s/\s+\z/\n/s;

    return $p->{result};
}

sub _textify_tag {
    my ($p, $tag, $skipped, $attrs) = @_;

    return if $p->{inside_bad_element};

    $skipped =~ s/\s+/ /sg; # Multiple spaces,newlines in html is simple space
    $p->{result} .= _textify_decode($skipped);

    if ($tag eq 'script' || $tag eq 'style') {
        $p->{inside_bad_element} = $tag;
        return;
    }

    if ($tag eq 'br' || $tag eq 'p' || $tag eq 'div') {
        $p->{result} .= "\n" if $p->{result} ne q{};

    } elsif ($tag eq 'hr') {
        $p->{result} .= "\n" . '-' x 50 . "\n";

    } elsif ($tag eq 'a') {
        $p->{a_href} = $attrs->{href}; # Save href for comparison with text content
        $p->{a_href_start} = length($p->{result});
    }
}

sub _textify_end_tag {
    my ($p, $tag, $skipped) = @_;

    unless ($p->{inside_bad_element}) {
        $skipped =~ s/\s+/ /sg; # Multiple spaces,newlines in html is simple space
        $p->{result} .= _textify_decode($skipped);
    } elsif ($p->{inside_bad_element} eq $tag) {
        $p->{inside_bad_element} = undef
    }

    # '<a href="URL">here</a>' => 'here [URL]'
    if ($tag eq 'a' && $p->{a_href}) {
        my $href    = $p->{a_href};
        $href =~ s{^(?:
                http:// (?=www\.) |
                ftp://  (?=ftp\.) |
                mailto:
            ) }{}xi;
        $href =~ s{/$}{};

        if (substr($p->{result}, $p->{a_href_start}) !~ /\Q$href\E/i) { # Content not contail URL?
            $p->{result} .= (substr($p->{result}, -1) eq ' ' ? '' : ' ') .
                '[' . $p->{a_href} . ']';
        }
        undef $p->{a_href}, $p->{a_href_start};
    }
}

sub _textify_decode {
    my $t = unescapeHTMLFull($_[0]);
    $t =~ tr/\xA0/ /; # nbsp => space
    return $t;
}

1;


lib/Mail/ToAPI/Checkvist.pm
package Mail::ToAPI::Checkvist;

use strict;
use warnings;
use 5.010;
our $VERSION = '0.4';

use WebService::Simple;
use Email::MIME;
use Email::Address;
use List::Util qw/first/;
use URI;
use Encode;

use Mail::ToAPI::Text qw/_parse_for_content/;

our $API_Endpoint = 'http://checkvist.com';
our $Chv;
our $Last_Error;

sub _init_api {
    my ($login, $remotekey) = @_;
    my $uri = URI->new($API_Endpoint);

    unless ($Chv) {
        $Chv = WebService::Simple->new(
            base_url        => $API_Endpoint,
            response_parser => 'JSON',
        );

        $Chv->default_headers->authorization_basic($login, $remotekey);
    }

    return $Chv;
}

sub _parse_to {
    my $to = shift;
    my ($login, $remotekey, $list_id, $list_tag);

    $to   = (Email::Address->parse($to))[0]->user;

    # = given, but I need "redo"
    for ($to) {
        s/^ post \+//xms;

        # we do not allow "." in remotekey to distinguish between
        # user$domain.com+remotekey@ and
        # remotekey+list_tag
        # fortunately, Checkvist remotekeys do not contain "."
        when (/^ ([^+.]+) \+ (\d+) $/xms) {
            ($remotekey, $list_id) = ($1, $2);
        }
        when (/^ ([^+.]+) \+ ([^+]+) $/xms) {
            ($remotekey, $list_tag) = ($1, $2);
        }
        when (/^ ([^+.]+) $/xms) {
            $remotekey = $1;
        }

        if (s/^ ([^+]+) \+ //xms) {
            $login = $1;
            $login =~ tr/$=/@@/;

            redo;
        }
    }

    return ($login, $remotekey, $list_id, $list_tag);
}

sub parse_email {
    my $fh = pop;

    my ($from, $to, $subject);

    binmode $fh, ':bytes';
    my $email = Email::MIME->new(join "", <$fh>);

    $subject = $email->header('Subject');
    $from = (Email::Address->parse($email->header_obj->header_raw('From')))[0]->address;

    my ($login, $remotekey, $list_id, $list_tag) = _parse_to($email->header_obj->header_raw('To'));

    my ($body_text, $files)   = _parse_for_content($email);

    $Last_Error = '';
    return {
        type        => 'add_task',

        login       => $login // $from,
        remotekey   => $remotekey,

        list_id     => $list_id,
        list_tag    => $list_tag,
        text        => $subject,

        ($body_text
            ? (note         => $body_text)
            : ()),
        ($files && @$files
            ? (files        => $files)
            : ()),
    };
}

sub execute {
    my $job = pop;

    my $rv;

    eval {
        given ($job->{type}) {
            when ('add_task') {
                $rv = _add_task($job);
            }
        }
    };

    if ($@) {
        $Last_Error = $@;
        undef $rv;
    }
    else {
        $Last_Error = '';
    }

    return $rv;
}

sub _add_task {
    my $job = shift;

    my $chv = _init_api($job->{login}, $job->{remotekey});

    unless ($job->{list_id}) {
        my $list_tag = $job->{list_tag} // 'inbox';   # default
        $job->{list_id} = "tag:$list_tag"; # supported by API
    }

    my %post_params = (
        Content => [
            import_content  => encode_utf8($job->{text}),
            parse_tasks     => 1,
        ],
    );

    if ($job->{note}) {
        push @{$post_params{Content}},
            import_content_note => encode_utf8($job->{note});
    }

    # in: [ [$filename, $content_type, $data] ... ]
    # out: [ undef, $filename,
    #   Content_Type => $content_type, Content => $data ]
    if ($job->{files}) {
        $post_params{Content_Type} = 'form-data';

        my $num = 1;
        for my $file (@{$job->{files}}) {
            my $ct = Mail::ToAPI::Text::_parse_header_fields("ct=$file->[1]");
            push @{$post_params{Content}},
                "add_files[$num]"   => [
                    undef,
                    encode_utf8($file->[0]),    # XXX
                    Content_Type    => $file->[1],
                    Content         => ($ct->{charset} ? encode($ct->{charset}, $file->[2]) : $file->[2]),
                ];
            ++$num;
        }
    }

    $Last_Error = '';
    my $rv = $chv->post(
        "checklists/$job->{list_id}/import.json",
        %post_params
    );

    if ($rv->is_success) {
        return $rv->parse_response->[0];
    }
    else {
        $Last_Error = $rv->status_line;
        return;
    }
}

sub fetch_tasks {
    my ($login, $remotekey, $list_id) = @_;
    my $chv = _init_api($login, $remotekey);

    $Last_Error = '';
    my $rv = $chv->get("checklists/$list_id/tasks.json",
        { with_notes    => 1 });

    if ($rv->is_success) {
        return $rv->parse_response;
    }
    else {
        $Last_Error = $rv->status_line;
        return;
    }
}

sub last_error {
    return $Last_Error;
}

1;
__END__

=encoding utf-8

=head1 NAME

Mail::ToAPI::Checkvist - Email in tasks for Checkvist.com, see Mail2CV.com

=head1 SYNOPSIS

  use Mail::ToAPI::Checkvist;

or

  % mail2cv.pl < email.eml

=head1 DESCRIPTION

This is the script that powers Mail2CV.com.

It is a service that allows adding tasks to your Checkvist.com list via email.

You will need a fairly modern perl and some modules from CPAN:

uni::perl, WebService::Simple, Email::MIME, Email::Address

You will also need to set up some email forwarding to feed incoming
emails to the script.  I use a catch-all virtual domain in Postfix and
a simple ~/.forward with a pipe.

=head1 AUTHOR

Alex Kapranoff E<lt>alex@kapranoff.ruE<gt>

=head1 COPYRIGHT

Copyright 2013 Alex Kapranoff

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License, Version
3.

=head1 SEE ALSO

L<http://checkvist.com>

=cut
